Add common checkout terms component

daledupreez opened the pull request:
#### Changes proposed in this Pull Request

* This PR adds a new `CheckoutTermsItem` component to provide more consistent behaviour for terms that we may need to show in checkout.
* Even though `client/my-sites/checkout/composite-checkout` is separate from `@automattic/composite-checkout`, there seems to be a practice of using CSS in JS in both places. Therefore, I moved the styles for the `CheckoutTermsItem` component into the JS component itself.
* I minimised the amount of styling for the checkout terms in `client/my-sites/checkout/upsell-nudge/purchase-modal/style.scss`. To be clear, we still need that little bit of CSS here to preserve the differences in styling between the checkout page and the upsell nudge.
* I removed unused CSS classes for checkout terms items. Since we're standardising the styles for this new component, it makes sense to not unnecessarily use separate (and unstyled) CSS classes for the same component.

#### Testing instructions

Note that the `CheckoutTermsItem` component is used for other products than just domains. There seems to be only two visually distinct scenarios where terms are displayed however: the checkout page and the upsell nudge. Therefore, my thinking is that it should suffice to test both of these scenarios.

To test the changes on the **checkout page**:
1. Add e.g. a domain to your cart.
1. Proceed to /checkout.

To test the changes in the **upsell nudge**:
1. Have a connected custom domain
1. Go to `/purchases/billing-history/{site-slug}`
1. Find a domain purchase, open it, copy the receipt ID
1. Go to `/checkout/offer-professional-email/{domain}/{receiptId}/{siteSlug}`
1. Fill out the input fields and click "Add professional email"

#### Screenshots of the changes

<img width="1419" alt="Screenshot 2022-04-28 at 15 35 48" src="https://user-images.githubusercontent.com/1101677/166216156-67ba1c36-39cf-455e-9cb6-5de22bd4c726.png">
<img width="1426" alt="Screenshot 2022-04-28 at 16 31 15" src="https://user-images.githubusercontent.com/1101677/166216163-264b8377-6fd4-412a-a9c0-53c81871ccae.png">


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 1200182182542585-as-1201900052374140

daledupreez committed:
Add new CheckoutTermsItem component

daledupreez committed:
Use CheckoutTermsItem component in existing terms components

fredrikekelund committed:
Self contained styling for CheckoutTermsItem component

fredrikekelund committed:
Removed unused CSS classes for CheckoutTermsItem component

sirbrillig commented:
> Even though client/my-sites/checkout/composite-checkout is separate from @automattic/composite-checkout, there seems to be a practice of using CSS in JS in both places. 

You're totally right. Just for the record this was for two reasons, and although using emotion is very convenient (and I personally like its specificity), I don't know if they are ultimately good ones:

- Since `@automattic/composite-checkout` was built to be used inside and outside calypso, we needed a way to inject styles from an npm package and this worked well. When we were rebuilding calypso checkout and developing this package, this allowed us to use the same technique in both places.
- Over the last few years, many components have migrated between the npm package and calypso as the API stabilized, and it was easier to leave the styles as-is.

sirbrillig commented on client/my-sites/checkout/composite-checkout/components/checkout-terms-item.tsx:
```diff
@@ -0,0 +1,53 @@
+	children,
+	className,
+	onClick,
+	wrapChildren = true,
```
Neat solution, although an optional boolean that defaults to true feels like a less-than-ideal API to me. What about renaming it to `noWrapChildren` or something and then defaulting to false? That way you can opt-out (which ultimately is what this does) by just specifying the bare prop like this:

```js 
<CheckoutTermsItem noWrapChildren>...</CheckoutTermsItem>
```

which feels more semantic to me than:

```js 
<CheckoutTermsItem wrapChildren={false}>...</CheckoutTermsItem>
```

stephanethomas replied:
I see your points but I'm on the fence. I'm not a fan of `noWrapChildren` because it introduces a negation, which makes its name a little bit more difficult to reason about than just `wrapChildren` ([this post](https://davej.io/2018/05/negative-naming.html) better explains this). Another problem with this approach is that it will force us to add `noWrapChildren` in a lot of places whereas we would need to declare `wrapChildren` once. Maybe `withoutParagraph` could be a good middle-ground 😀?

fredrikekelund replied:
Fair point! I went with `prewrappedChildren` and inverted the truthy/falsy logic. WDYT, is that an acceptable prop name?

fredrikekelund replied:
I should also mention that I looked at the possibility of removing this prop entirely in favor of inspecting the `children` to see if they contain block elements or not. Turns out that this very quickly gets complex once you have block elements nested within `Fragment`s (which we have), so I abandoned this approach.

sirbrillig replied:
> I'm not a fan of noWrapChildren because it introduces a negation, which makes its name a little bit more difficult to reason about than just wrapChildren

I agree! If `wrapChildren` defaulted to false, then I think it would be better, but I was trying to come up with a way to keep wrapping as the default without having to have a truthy optional prop default.

> Fair point! I went with prewrappedChildren and inverted the truthy/falsy logic. WDYT, is that an acceptable prop name

👍  Sounds fine to me. Ultimately someone using this component needs to understand what "wrapped" means anyway, but at least this way there might be one fewer surprises!

> I should also mention that I looked at the possibility of removing this prop entirely in favor of inspecting the children to see if they contain block elements or not. Turns out that this very quickly gets complex once you have block elements nested within Fragments (which we have), so I abandoned this approach.

Yes, learning about children is very tricky. [It can be done](https://github.com/Automattic/wp-calypso/blob/41b943648e2f9b45000af1706e843ba7212b235c/packages/composite-checkout/src/components/checkout-steps.tsx#L162-L163) but I've had the same problems you describe here. Avoiding it as you've done is a good idea.

stephanethomas replied:
Since `prewrappedChildren` is a boolean, I think I would go with `isPrewrappedChildren`.

fredrikekelund replied:
Done! I've renamed the prop

stephanethomas commented on client/my-sites/checkout/composite-checkout/components/checkout-terms-item.tsx:
```diff
@@ -0,0 +1,53 @@
+const CheckoutTermsItem = ( {
+	children,
+	className,
+	onClick,
```
I think we should define a default value, i.e. a `noop` function.

stephanethomas replied:
Or maybe not after all. I thought this would trigger an error if I clicked the terms where that even handler wouldn't be provided but I didn't notice anything in my browser's console.

fredrikekelund replied:
We should be fine without a noop default! React accepts `undefined` or a function as the `onClick` prop to a `div`. But just for the sake of it – I updated the type for `onClick` to be a little bit more strict.

stephanethomas reviewed (approved):
Very good job with the description of this pull request! I only had minor comments which don't have to be addressed necessarily (I'll let you decide).

<img width="619" alt="screenshot" src="https://user-images.githubusercontent.com/594356/166426953-7bb52135-d3fa-4e4b-8639-1cf8cade4ae0.png">

![image](https://user-images.githubusercontent.com/594356/166427045-9e6acb3d-e21a-441a-922f-4c3ef9963bfc.png)

And thanks @sirbrillig for reviewing it!

AllTerrainDeveloper commented:
One nit! Although PR description is awesome, there is a missing link:

Related to #

We usually link it to our Asana Task using the TamperMonkey script, which at the end results in something like: {number-as-number2} which refers to our Asana Card.

fredrikekelund committed:
Inverted wrapChildren prop logic

Pass a truthy prop to avoid wrapping children of `CheckoutTermsItem` instead of passing a falsy prop to do the same

stephanethomas commented:
> We usually link it to our Asana Task

Is this a new convention? I've never done that, and wonder if that's really useful ...

fredrikekelund committed:
Renamed prewrappedChildren prop

Also removed the presentation ARIA role from the CheckoutTermsItem component

fredrikekelund commented:
I also removed the ARIA `presentation` role from the div wrapper in `CheckoutTermsItem`. That role should not necessarily obscure the content of the div from screen readers ([ref](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/presentation_role)), but `div`s already have an implicitly generic semantic purpose and given that the checkout terms are important for users to understand, I don't see any reason for reducing the importance of the wrapper div.

AllTerrainDeveloper commented:
> > We usually link it to our Asana Task
> 
> Is this a new convention? I've never done that, and wonder if that's really useful ...

I'm reviewing changes and I've just seen your comment after our Meetup! Ups, I said we, but probably is mostly only myself? I find it useful to check the asana card status
